### PR TITLE
Scheduler V2 Bugfixes

### PR DESCRIPTION
**Pull request by jzila**

### CODE REVIEW

**Comment by jzila** [comment]
The ControllerSuite and SchedulerSuite tests run successfully on my dev machine; not sure what's going on in CI.

**Comment by jzila** [comment]
Starting to look a lot better: https://ci.flynn.io/builds/20150909070756-14914cf0

File: controller/scheduler-v2/scheduler.go [hunk 1]
```diff
@@ -498,7 +498,7 @@ func (s *Scheduler) HandleJobRequest(req *JobRequest) {
 func (s *Scheduler) RunPutJobs() {
 	log := logger.New("fn", "RunPutJobs")
 	log.Info("starting job persistence loop")
-	strategy := attempt.Strategy{Delay: 100 * time.Millisecond, Total: time.Minute}
+	strategy := attempt.Strategy{Delay: 100 * time.Millisecond, Total: time.Second}
```
Comment (lmars) [inline controller/scheduler-v2/scheduler.go]: The controller client [uses a retry dialer](https://github.com/flynn/flynn/blob/scheduler-v2-review/controller/client/client.go#L134) which use a [1s connect timeout and 30s retry timeout](https://github.com/flynn/flynn/blob/scheduler-v2-review/pkg/dialer/retry.go#L10-L20), so having a 1s strategy here probably doesn't make sense.

It is actually quite important for these calls to eventually succeed (see #686), so maybe we need a pool of goroutines which are calling `PutJob`? We need to make sure they are not delivered out of order for any given job though (see [this DB constraint](https://github.com/flynn/flynn/blob/master/controller/schema.go#L132-L140)).

File: controller/scheduler-v2/job.go [hunk 2]
```diff
@@ -4,6 +4,7 @@ import (
 	"time"
 
 	ct "github.com/flynn/flynn/controller/types"
+	host "github.com/flynn/flynn/host/types"
```
Comment (lmars) [inline controller/scheduler-v2/job.go]: This package is already called `host`

File: controller/scheduler-v2/scheduler.go [hunk 3]
```diff
@@ -913,18 +919,18 @@ func (s *Scheduler) Unsubscribe(events chan Event) {
...
 		fallthrough
 	case host.StatusRunning:
 		s.handleJobStart(job)
-		controllerState = "up"
 	default:
 		delete(s.jobs, job.JobID)
 		delete(s.stoppedJobs, job.JobID)
 	}
-	s.putJobs <- controllerJobFromSchedulerJob(job, controllerState, metadata)
+	log.Info("Queuing job for persistence")
```
Comment (lmars) [inline controller/scheduler-v2/scheduler.go]: Log messages should not be capitalised

**Comment by jzila** [comment]
@lmars please disregard the last commmit: I figured out a better approach to achieve the same goal.

File: controller/scheduler-v2/scheduler.go [hunk 4]
```diff
@@ -18,25 +20,40 @@ import (
...
-	serviceName         = "controller-scheduler"
+	eventBufferSize    int           = 1000
+	serviceName                      = "controller-scheduler"
+	maxJobAttempts     uint          = 30
+	jobAttemptInterval time.Duration = 500 * time.Millisecond
 )
 
 var logger = log15.New("component", "scheduler")
 
+func fnLogger(ctx ...interface{}) log15.Logger {
```
Comment (jzila) [inline controller/scheduler-v2/scheduler.go]: I added this to simplify searching (both in the editor and grep). This way function names only show up once.

File: controller/scheduler-v2/scheduler.go [hunk 5]
```diff
@@ -84,10 +101,12 @@ func NewScheduler(cluster utils.ClusterClient, cc utils.ControllerClient) *Sched
 		jobRequests:      make(chan *JobRequest, eventBufferSize),
 		putJobs:          make(chan *ct.Job, eventBufferSize),
 	}
+	s.maxBackoffPeriod = 10 * s.backoffPeriod
```
Comment (jzila) [inline controller/scheduler-v2/scheduler.go]: Fix for #1340

File: controller/scheduler-v2/scheduler.go [hunk 6]
```diff
@@ -736,26 +763,51 @@ func (s *Scheduler) updateFormation(f *ct.Formation) (*Formation, error) {
...
+	newReq := &JobRequest{
+		RequestType: req.RequestType,
+		Job:         &job,
+	}
 	defer func() {
+		s.pendingStarts.AddJob(newReq.Job)
 		if err != nil {
-			s.scheduleJobRequest(req)
 			log.Error("error starting job", "err", err)
+			if newReq.attempts < maxJobAttempts {
```
Comment (jzila) [inline controller/scheduler-v2/scheduler.go]: Limit number of times a failed job request can try to start. Rely on `Rectify` to handle the rest.

File: controller/scheduler-v2/scheduler.go [hunk 7]
```diff
@@ -936,49 +995,73 @@ func (s *Scheduler) Jobs() map[string]*Job {
...
 	log.Info("adding job to in-memory state")
 	_, ok := s.jobs[job.JobID]
 	if s.isLeader && !ok && s.pendingStarts.HasStarts(job) {
 		s.pendingStarts.RemoveJob(job)
 	}
 	s.jobs[job.JobID] = job
+	delete(s.stoppedJobs, job.JobID)
+}
+
+func (s *Scheduler) handleJobCrash(job *Job) {
```
Comment (jzila) [inline controller/scheduler-v2/scheduler.go]: Implemented restarting of crashed jobs.

File: controller/scheduler-v2/scheduler_test.go [hunk 8]
```diff
@@ -282,13 +282,13 @@ func (TestSuite) TestExponentialBackoffNoHosts(c *C) {
 			c.Fatalf("expected JobRequestEvent, got %T", event)
 		}
 		c.Assert(e.Request, NotNil)
-		c.Assert(e.Request.restarts, Equals, expected)
 	}
 
 	// wait for the formation to cascade to the scheduler
 	s.waitRectify()
-	waitRestarts(1*time.Second+50*time.Millisecond, 2)
-	waitRestarts(2*time.Second+50*time.Millisecond, 3)
```
Comment (jzila) [inline controller/scheduler-v2/scheduler_test.go]: Switch to linear restart attempts for job requests (exponential for crashed job restarts)

File: controller/scheduler-v2/scheduler.go [hunk 9]
```diff
@@ -936,49 +995,73 @@ func (s *Scheduler) Jobs() map[string]*Job {
...
+	log := fnLogger()
+	if req.startedAt.Before(time.Now().Add(-s.backoffPeriod)) {
+		log.Info("resetting job restarts", "backoffPeriod", s.backoffPeriod, "job.startedAt", req.startedAt)
+		req.restarts = 0
 	}
-}
-
-func (s *Scheduler) scheduleJobRequest(req *JobRequest) {
 	backoff := s.getBackoffDuration(req.restarts)
+	if backoff == s.maxBackoffPeriod {
```
Comment (jzila) [inline controller/scheduler-v2/scheduler.go]: Don't use exponential backoff past the max period (`10 * s.backoffPeriod`). After that, the attempt will fail and we'll have to rely on `Rectify`.

**Comment by lmars** [comment]
@jzila the unit tests are failing with:

```
controller/scheduler/main.go:939: too many arguments in call to utils.JobConfig
```

**Comment by jzila** [comment]
Ah, sry about that. I wonder why the incremental `make` on my machine didn't pick that up...

**Comment by lmars** [comment]
@jzila we are no longer building scheduler v1 on this branch, we should probably just delete it

**Comment by jzila** [comment]
Looks like when a release is added to the `controller` app, the scheduler correctly stops itself upon which the controller starts a new one. However, we don't notify the controller with a final `PutJob`. In fact, I'm unclear as to how the controller is starting a new scheduler given that it isn't notified about it via PutJob until the next scheduler comes up.

Maybe we should add in a `SIGTERM` handler, upon which we write to the `s.stop` channel instead of terminating right away.

**Comment by lmars** [comment]
@jzila 

> when a release is added to the controller app, the scheduler correctly stops itself upon which the controller starts a new one

This is incorrect. A new controller release leads to a one-by-one deployment, so the deployer incrementally scales up the new release and scales down the old release. Only the schedulers are actually starting / stopping jobs, and when the current leader is scaled down, one of the new schedulers should take the reign and continue reacting to the scaling being done by the deployer.

> we don't notify the controller with a final PutJob

True, but the other schedulers should see the leader go down and call PutJob themselves. This actually highlights the fact that when the scheduler starts, it needs to wait until it is watching all the hosts *before* registering with discoverd, so the deployer doesn't scale things too fast and lead to no schedulers watching all hosts.

> Maybe we should add in a SIGTERM handler, upon which we write to the s.stop channel instead of terminating right away.

Makes sense, we can do this easily using `shutdown.BeforeExit`.

**Comment by lmars** [comment]
@jzila the latest build is showing ~100 schedulers running on host `a0a72333` :scream:

I am looking into it, and I may make some code changes. If I do, I will open a PR with my changes on top of yours (I can't push to this PR).

**Comment by lmars** [comment]
@jzila I am going to merge this then I have a few fixes I will PR separately

**Pull request merged**